### PR TITLE
Add dialog to enter status text

### PR DESCRIPTION
**Pull request by mightymop**

### CODE REVIEW

File: src/ui/Roster.ts [hunk 1]
```diff
@@ -359,19 +360,26 @@ export default class Roster {
...
+            showSetExStatusDialog();
          }
-
-         if (oldPresence === Presence.offline && requestedPresence === Presence.online) {
-            let onUserRequestsToGoOnline = Client.getOption('onUserRequestsToGoOnline');
-
-            if (typeof onUserRequestsToGoOnline === 'function') {
-               onUserRequestsToGoOnline();
-            }
+         else
```
Comment (sualko) [inline src/ui/Roster.ts:367]: Returning inside the if clause would help to reduce indention and enhance the diff too.

File: src/ui/dialogs/exstatus.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   dom.find('form').submit(onSubmit);
+
+   $('.jsxc-js-clear').on('click',removeStatus);
+}
+
+function onSubmit()
+{
+   let account = Client.getAccountManager().getAccounts()[0];
+   let statustext = dialog.getDom().find('#extended_status_text').val().toString();
+   let item = $build('status',{xmlns:'http://jabber.org/protocol/status'}).c('text').t(statustext).tree();
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:29]: We should move this to a service.

File: src/ui/dialogs/exstatus.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+   dom.find('form').submit(onSubmit);
+
+   $('.jsxc-js-clear').on('click',removeStatus);
+}
+
+function onSubmit()
+{
+   let account = Client.getAccountManager().getAccounts()[0];
+   let statustext = dialog.getDom().find('#extended_status_text').val().toString();
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:28]: Use e.g. form name instead of id.

File: src/ui/dialogs/exstatus.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import Dialog from '../Dialog'
+import Client from '../../Client'
+//import JID from '../../JID'
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:3]: Should be removed.

File: src/ui/dialogs/exstatus.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+function onSubmit()
+{
+   let account = Client.getAccountManager().getAccounts()[0];
+   let statustext = dialog.getDom().find('#extended_status_text').val().toString();
+   let item = $build('status',{xmlns:'http://jabber.org/protocol/status'}).c('text').t(statustext).tree();
+
+   account.getContact().setStatus(statustext);
+
+   account.getConnection().getPEPService().publish('http://jabber.org/protocol/status',item,'current').then(function(result) {
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:33]: In which XEP is this defined?

File: template/roster.hbs [hunk 6]
```diff
@@ -5,29 +5,33 @@
...
-        classes="jsxc-menu--pushup jsxc-grow jsxc-js-presence-menu"
-        button-label=(t "Offline")}}
-            <li data-presence="online" class="jsxc-online jsxc-offline-available">{{t "Online"}}</li>
-            <li data-presence="chat" class="jsxc-chat">{{t "Chatty"}}</li>
-            <li data-presence="away" class="jsxc-away">{{t "Away"}}</li>
-            <li data-presence="xa" class="jsxc-xa">{{t "Extended_away"}}</li>
-            <li data-presence="dnd" class="jsxc-dnd">{{t "dnd"}}</li>
-            <li data-presence="offline" class="jsxc-offline">{{t "Offline"}}</li>
-      {{/menu}}
+   <div class="jsxc-real-bottom">
```
Comment (sualko) [inline template/roster.hbs:8]: Real bottom was probably not intentionally included in this pr.

**Review by sualko — CHANGES_REQUESTED** [review]

File: src/ui/Roster.ts (hunk 1 above)
Comment (mightymop) [reply src/ui/Roster.ts]: done

File: src/ui/dialogs/exstatus.ts (hunk 3 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: ok i have moved the build code for the packets to pep service

File: src/ui/dialogs/exstatus.ts (hunk 4 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: removed

File: src/ui/dialogs/exstatus.ts (hunk 5 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: oh i now see it is not official... so we could remove it or not... up to you...
found it here: http://www.funkroom.net/xep/xep-status.html

File: template/roster.hbs (hunk 6 above)
Comment (mightymop) [reply template/roster.hbs:8]: right... C&P error ;D

File: src/ui/dialogs/exstatus.ts (hunk 2 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: ok i have moved the build code for the packets to pep service

File: src/connection/services/PEP.ts [hunk 7]
```diff
@@ -23,6 +23,16 @@ export default class PEP extends AbstractService {
       }
    }
 
+   public publisStatus(statustext:string)
```
Comment (sualko) [inline src/connection/services/PEP.ts:26]: Typo

File: src/ui/dialogs/exstatus.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+   });
+
+   $('.jsxc-js-save').on('click',sendStatus);
+   $('.jsxc-js-clear').on('click',removeStatus);
+}
+
+function sendStatus()
+{
+
+   let account = Client.getAccountManager().getAccounts()[0];
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:26]: We should loop through all accounts

File: src/ui/dialogs/exstatus.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+let dialog: Dialog;
+
+export default function() {
+
+   let content = exstatusTemplate();
+
+   dialog = new Dialog(content);
+   let dom = dialog.open();
+
+   $(dom).ready(function(){
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:15]: No need to wait

File: src/ui/dialogs/exstatus.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+   account.getConnection().getPEPService().publisStatus(statustext).then(function(result) {
+         dialog.close();
+   }).catch(function (msg){
+         dialog.close();
+   });
+}
+
+function getStatus()
+{
+     let account = Client.getAccountManager().getAccounts()[0];
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:43]: We should check all accounts and than take the first one.

File: src/ui/dialogs/exstatus.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+     let jid = contact.getJid();
+     account.getConnection().getPEPService().retrieveItems('http://jabber.org/protocol/status',jid.bare).then(function(data) {
+         dialog.getDom().find('textarea[name="extended_status_text"]').val($(data).text());
+     });
+}
+
+function removeStatus(ev) {
+   ev.preventDefault();
+
+   let account = Client.getAccountManager().getAccounts()[0];
```
Comment (sualko) [inline src/ui/dialogs/exstatus.ts:54]: Loop

**Review by sualko — CHANGES_REQUESTED** [review]
In my opinion we should not implement http://www.funkroom.net/xep/xep-status.html, because it's lacking crucial information. The biggest issue is, that it's not configuring a single item node. So we would probably end up with multiple old status messages. No big deal to fix it (see omemo), but I don't see a need for it either.

Otherwise looks good.

File: src/connection/services/PEP.ts (hunk 7 above)
Comment (mightymop) [reply src/connection/services/PEP.ts]: removed...

File: src/ui/dialogs/exstatus.ts (hunk 9 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: changed...

File: src/ui/dialogs/exstatus.ts (hunk 11 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: added account choice

**Comment by mightymop** [comment]
> 
> 
> In my opinion we should not implement http://www.funkroom.net/xep/xep-status.html, because it's lacking crucial information. The biggest issue is, that it's not configuring a single item node. So we would probably end up with multiple old status messages. No big deal to fix it (see omemo), but I don't see a need for it either.
> 
> Otherwise looks good.

i removed that unofficial xep..

File: src/ui/dialogs/exstatus.ts (hunk 8 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: added account choice

File: src/ui/dialogs/exstatus.ts (hunk 10 above)
Comment (mightymop) [reply src/ui/dialogs/exstatus.ts]: added account choice

**Comment by sualko** [comment]
I merged it manually, because I changed some things:

- I removed the account and presence selection, because this would result in confusion. The roster bottom can only show one presence, so it would not be clear which account is using which presence.
- Status can now be added to every presence.
- The status is shown in the bottom part of the roster.

image

**Comment by mightymop** [comment]
> 
> 
> I merged it manually, because I changed some things:
> 
>     * I removed the account and presence selection, because this would result in confusion. The roster bottom can only show one presence, so it would not be clear which account is using which presence.
> 
>     * Status can now be added to every presence.
> 
>     * The status is shown in the bottom part of the roster.
> 
> 
> image

very nice improvements...

**Pull request closed without merging**